keras/engine/training.py

```diff
@@ -1150,6 +1164,10 @@ def _fit_loop(self, f, ins, out_labels=None, batch_size=None,
                 indices_for_conversion_to_dense.append(i)
 
         for epoch in range(initial_epoch, epochs):
+            # Reset stateful metrics
+            for m in self.metrics:
+                if isinstance(m, Layer):
+                    m.reset_states()
```

Comment:
Wouldn't this lead to cryptic error if someone forgets to implement reset_states?

The only assumption here is that m is a Layer. And there is no way of knowing what is a stateful metric. I would like to see a compromise between your approach and brge's approach. 

```
class StatefulMetric(Layer):
    def reset_states():
        raise NotImplementedError
```

Reply:
As a user, what can I do, knowing that I can now feed a Layer to the `metrics` arguments? Can I feed any Layer? (No, but some will try)

Reply:
Good point

Reply:
I prefer the middle ground as @Dref360 mentioned.

But, I'm fine either way.